File: Pod/Classes/ViewObject/SSAgentViewObject.m
//
// Created by P.I.akura on 2013/08/18.
// Copyright (c) 2013 P.I.akura. All rights reserved.
//
// To change the template use AppCode | Preferences | File Templates.
//


#import "SSAgentViewObject.h"
#import "TableViewAgentProtocol.h"

@implementation SSAgentViewObject

- (id)initWithArray:(NSArray *)array {
    self = [super init];
    if (self) {
        _array = array.mutableCopy;
    }
    return self;
}

- (void)addObject:(id)object {
    [_array addObject:object];
    [self.agent insertCell:self atIndexPath:[NSIndexPath indexPathForItem:_array.count - 1 inSection:0]];
}

- (void)changeObject:(id)object {
    NSIndexPath *path = [self indexPathForObject:object];
    if (path) {
        [self.agent changeUpdateCell:self atIndexPath:path];
    }
}

- (void)removeObjectAtIndexPath:(NSIndexPath *)indexPath {
    [_array removeObjectAtIndex:indexPath.row];
    [self.agent deleteCell:self atIndexPath:indexPath];
}

- (NSIndexPath *)indexPathForObject:(id)object {
    for (NSInteger i = 0, _len = _array.count; i < _len; i++) {
        if ([_array[i] isEqual:object]) {
            return [NSIndexPath indexPathForItem:i inSection:0];
        }
    }
    return nil;
}

- (void)dealloc {
    _array = nil;
}

- (NSUInteger)countInSection:(NSUInteger)section {
    return _array.count;
}

- (id)objectAtIndexPath:(NSIndexPath *)indexPath {
    id object = _array[indexPath.row];
    return self.convert ? self.convert(object) : object;
}

- (NSArray *)sectionObjects:(NSInteger)section {
    return _array;
}

- (NSUInteger)sectionCount {
    return 0 < _array.count;
}

- (void)setEditing:(BOOL)editing {
}
@end

File: lib/ViewObject/SSAgentViewObject.swift
//
//  SSAgentViewObject.swift
//  TableViewAgent
//
//  Created by akuraru on 2014/07/29.
//  Copyright (c) 2014年 P.I.akura. All rights reserved.
//

import Foundation
import UIKit

class SSAgentViewObject<T: NSObject, U: NSObject> : AgentViewObject<T, U> {
    weak var agent :TableViewAgent<T, U>!
    var array: [T]
    
    init(array :[T], agent: TableViewAgent<T, U>, transform: T -> U = ({t in t} as T -> U)) {
        self.agent = agent
        self.array = array
        super.init(transform)
    }
    func indexPathForObject(object :T) -> NSIndexPath? {
         let _len  = array.count
        for var i = 0; i < _len; i++ {
            if object.isEqual(array[0]) {
                return NSIndexPath(forRow: i, inSection: 0)
            }
        }
        return nil
    }
    override func addObject(object :T,inSection section :Int) {
        array.append(object)
        agent.insertCell(NSIndexPath(forRow: array.count - 1, inSection: 0))
    }
    override func changeObject(object :T) {
        let path = self.indexPathForObject(object)
        if path != nil {
            agent.changeUpdateCell(path!)
        }
    }
    // deleagte
    override func sectionCount() -> Int {
        return array.isEmpty ? 0 : 1
    }
    override func countInSection(section :Int) -> Int {
        return array.count
    }
    override func objectAtIndexPath(indexPath :NSIndexPath) -> U {
        return transform(array[indexPath.row])
    }
    override func removeObjectAtIndexPath(indexPath :NSIndexPath) {
        array.removeAtIndex(indexPath.row)
        agent.deleteCell(indexPath)
    }
    override func existObject(indexPath :NSIndexPath) -> Bool {
        let row = indexPath.row
        return 0 <= row && row < array.count
    }
    override func sectionObjects(section :Int) -> U {
        return transform(array[section])
    }
}